compiler/src/iree/compiler/Codegen/Common/GPU/GPUHeuristics.cpp

```diff
@@ -17,10 +20,88 @@ using llvm::APIntOps::GreatestCommonDivisor;
 
 namespace mlir::iree_compiler {
 
-std::optional<GPUMMASchedule>
+int64_t sharedMemoryUsed(const GPUMMASchedule &schedule, int64_t lhsBitwidth,
+                         int64_t rhsBitwidth) {
+  int64_t tileM = schedule.mSize * schedule.mTileCount * schedule.mWarpCount;
+  int64_t tileN = schedule.nSize * schedule.nTileCount * schedule.nWarpCount;
+  int64_t tileK = schedule.kSize * schedule.kTileCount;
+  return tileM * tileK * lhsBitwidth + tileN * tileK * rhsBitwidth;
+}
+
+bool isValidSchedule(const GPUMatmulShapeType &problem,
+                     const GPUMMASchedule &schedule) {
+  return problem.mSize %
+                 (schedule.mSize * schedule.mTileCount * schedule.mWarpCount) ==
+             0 &&
+         problem.nSize %
+                 (schedule.nSize * schedule.nTileCount * schedule.nWarpCount) ==
+             0 &&
+         problem.kSize % (schedule.kSize * schedule.kTileCount) == 0;
+}
+
+FailureOr<GPUMMASchedule>
+fitScheduleInSharedMemory(const GPUMatmulShapeType &problem,
+                          ArrayRef<GPUMatmulShapeType> intrinsics,
+                          GPUMMASchedule schedule, int64_t sharedMemLimit) {
+  int64_t lhsBitwidth =
+      intrinsics[schedule.index].aType.getIntOrFloatBitWidth() / 8;
+  int64_t rhsBitwidth =
+      intrinsics[schedule.index].bType.getIntOrFloatBitWidth() / 8;
+
+  while (!isValidSchedule(problem, schedule) ||
+         sharedMemoryUsed(schedule, lhsBitwidth, rhsBitwidth) >
+             sharedMemLimit) {
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "Shrinking schedule\n";
+      llvm::dbgs() << "mSize: " << schedule.mSize << "\n";
+      llvm::dbgs() << "nSize: " << schedule.nSize << "\n";
+      llvm::dbgs() << "kSize: " << schedule.kSize << "\n";
+      llvm::dbgs() << "mTileCount: " << schedule.mTileCount << "\n";
+      llvm::dbgs() << "nTileCount: " << schedule.nTileCount << "\n";
+      llvm::dbgs() << "kTileCount: " << schedule.kTileCount << "\n";
+      llvm::dbgs() << "mWarpCount: " << schedule.mWarpCount << "\n";
+      llvm::dbgs() << "nWarpCount: " << schedule.nWarpCount << "\n";
+    });
+
+    auto decrementIfPossible = [](int64_t &c) {
+      if (c <= 1) {
+        return false;
+      }
+      --c;
+      return true;
+    };
+
+    // Attempt to shrink the schedule along one of the dimensions.
+    // TODO: Is it worth trying to decrease only the tile count with the maximum
+    // size?
+    if (decrementIfPossible(schedule.mTileCount)) {
```

Comment:
The previous heuristic sort of avoided this, but this decrementing approach will essentially always decrement `mTileCount` to 1 before trying anything else. Beforehand it would find the smallest valid odd factor of `mTileCount` and then move on to decrementing `nTileCount` and so on. Both this and what was there before are borderline hacks (although this has better guarantees of fitting inside shared memory, so that's a plus). A better solution should probably factor `problem.mSize / (mWarpCount * mTileCount * mSize)` and then pop off the smallest factors one at a time, preferably trying to keep the tile "generally square."

Also it's likely easier to first shrink `mWarpCount * mTileCount` (i.e. the whole tile size) and then shrink the warp and batch counts separately.

Not blocking, just a note.

Reply:
Added a TODO about exploring better strategies for now :)